### DISCUSSION TITLE
Use TinyTuya to control Orvibo devices?

### DISCUSSION THREAD

**Discussion by Rhys-Baker** [jasonacox/tinytuya#113] [Q&A]
Hey everyone, I could use some help here. I recently purchased some roller shutters for some windows in my house. The company also sold me an Orvibo Allone hub. I would like to control them using python scripts. Does anyone know if there is a way / how to connect this device to the Tuya Smart app and, by extension, control them with python scripts.

If you know any forums to go to for help, that would be great, thank you!

**Comment by jasonacox** [comment]
Hi @Rhys-Baker - I looked at the Orvibo products and they appear to be *Zigbee devices* that use the "HomeMate" app to control them.  The TinyTuya project will only control Tuya based devices.  Devices that work with the "SmartLife" app will most likely work with TinyTuya.  

You might check out https://github.com/zigpy/zigpy.

**Reply by Rhys-Baker** [reply]
Thanks! I'll be sure to check that out

**Reply by J4CE2** [reply]
Did you get it working Rhys?

**Reply by Rhys-Baker** [reply]
No, I never really ended up looking at this again. If I get some time... ever, and I get bored enough I'll probably take a look into it.
I found other solutions for what I wanted it to do anyway.

If I end up looking into it, I'll be sure to post a solution here :)

**Comment by deepcoder** [comment]
Hi @jasonacox FYI, you might want to say ' control Tuya Wifi based devices', because there are Tuya zigbee devices as well. And they show up in the top level of the Tuya Smart app, even though they communicate via a wifi/zigbee bridge.

**Reply by jasonacox** [reply]
@deepcoder Thanks!  Do those zigbee bridge devices show up with TinyTuya scans (`python3 -m tinytuya scan`) or in the setup wizard?

**Reply by deepcoder** [reply]
Not that I have seen. I've not worked closely with the Tuya zigbee hubs. I just bought to check firmware versions of devices. These zigbee devices seem to use the same type DPS interface as their WiFi devices. Which is a totally different than the 10 bajillion zigbee devices that do a fairly good job of staying within the guard rails of the Zigbee specification.  I use the Tuya Zigbee devices with zigbee2mqtt with mixed results so far.

It would be interesting (time waster?) to see if there is some type of api to these gateways. Below is a simple test for one of the Zigbee Gateways. The Tuya mobile apps are able to communicate with these zigbee devices and they are doing this via WiFi :

```
For a Moes Zigbee WiFi Gateway here is what the iot.tuya.com query returns:

{
  "result": {
    "category": "wg2"
  },
  "success": true,
  "t": 1647660433991,
  "tid": "783d656fa73411ecba1a0276a031ac08"
}

-- here is what your program returns:

{
    "code": 2009,
    "msg": "not support this device",
    "success": false,
    "t": 1647660336710,
    "tid": "3e3febb8a73411eca6216e0e3d4c7b49"
}

```

**Reply by deepcoder** [reply]
Here is the is the output of the tinytuya scan and the parts of the tuya-raw.json for these devices :

Small Bedroom Luminance sensor is a zigbee battery powered end device
Moes Zigbee WiFi Gateway is the wifi to zigbee bridge


```
a@b:~/tinytuya$ python -m tinytuya wizard
TinyTuya Setup Wizard [1.3.1]

    Existing settings:
        API Key=x
        Secret=y
        DeviceID=z
        Region=us

    Use existing credentials (Y/n):


Device Listing

[
    {
        "name": "Moes Zigbee WiFi Gateway",
        "id": "x",
        "key": "y"
    },
.
.
.
.
.

    {
        "name": "Small Bedroom Luminance sensor",
        "id": "x",
        "key": "y"
    }
]

>> Saving list to devices.json
    10 registered devices saved

>> Saving raw TuyaPlatform response to tuya-raw.json

Poll local devices? (Y/n):

Scanning local network for Tuya devices (retry 35 times)...
    8 local devices discovered

Polling local devices...
    [Moes Zigbee WiFi Gateway] - 10.10.1.192 - No Response
    [Heater_Small_Bedroom] - 10.10.1.180 - Off - DPS: {'1': False, '2': 70, '3': 67, '4': 'high', '8': False, '10': True, '11': '0h', '12': 0, '102': False, '103': False}
    [Galaxy AI-Plus Breakfast Table] - 10.10.1.206 - On - DPS: {'20': True, '21': 'white', '22': 1000, '23': 481, '24': '00d800f00033', '25': '000e0d0000000000000000c80000', '26': 0}
    [Ustellar Tunable White Light Strip] - 10.10.1.203 - Off - DPS: {'20': False, '21': 'white', '22': 435, '23': 0, '25': '000e0d0000000000000000c80000', '26': 0}
    [Small Bedroom Sunbeam Bedding] - 10.10.1.201 - Off - DPS: {'13': 0, '14': False, '20': 'level_3', '24': False, '26': '10h', '28': 36000, '102': 0, '103': 'User_Confirmation'}
    [Galaxy AI-Plus Sink] - 10.10.1.184 - On - DPS: {'20': True, '21': 'white', '22': 1000, '23': 481, '24': '003c03e803e8', '25': '000e0d0000000000000000c80000', '26': 0}
    [Heater_Dining_Room] - 10.10.1.167 - Off - DPS: {'1': False, '2': 83, '3': 69, '4': 'low', '8': False, '10': True, '11': '0h', '12': 0, '102': False, '103': False}
    [100W Tune + Color] - 10.10.1.207 - Off - DPS: {'20': False, '21': 'scene', '22': 160, '23': 1000, '24': '00ff03e803e8', '25': '0032320000000000000001f40000', '26': 0, '41': True}
    [Flood-01] - 0 - Error: No IP found
    [Small Bedroom Luminance sensor] - 0 - Error: No IP found

>> Saving device snapshot data to snapshot.json

Done.

a@b:~/tinytuya$


--- relevant sections of tuya-raw.json
        {
            "active_time": 1645399327,
            "biz_type": 0,
            "category": "wg2",
            "create_time": 1645399236,
            "icon": "smart/icon/ay15327721968035jwx9/f3ee42e90b543f1c0a50279f8f450cad.png",
            "id": "x",
            "ip": "x",
            "lat": "x",
            "local_key": "x",
            "lon": "x",
            "model": "",
            "name": "Moes Zigbee WiFi Gateway",
            "online": true,
            "owner_id": "x",
            "product_id": "ysgynkrk",
            "product_name": "Gateway",
            "status": [],
            "sub": false,
            "time_zone": "-08:00",
            "uid": "x",
            "update_time": 1647580616,
            "uuid": "x"
        }


        {
            "active_time": 1645399943,
            "biz_type": 0,
            "category": "ldcg",
            "create_time": 1645399943,
            "icon": "smart/program_category_icon/ldcg.png",
            "id": "x",
            "ip": "",
            "lat": "x",
            "local_key": "x",
            "lon": "x",
            "model": "S-LUX-ZB",
            "name": "Small Bedroom Luminance sensor",
            "node_id": "x",
            "online": true,
            "owner_id": "x",
            "product_id": "pisltm67",
            "product_name": "Luminance sensor",
            "status": [
                {
                    "code": "bright_state",
                    "value": "low"
                },
                {
                    "code": "bright_value",
                    "value": 0
                },
                {
                    "code": "battery_percentage",
                    "value": 100
                }
            ],
            "sub": true,
            "time_zone": "-08:00",
            "uid": "x",
            "update_time": 1645399963,
            "uuid": "x"
        }
        
```

**Reply by jasonacox** [reply]
Thanks @deepcoder !   Some thoughts on this:
* I have seen gateways before and they typically don't offer anything in terms of DPS (however some have temp and humidity datapoints).  Likely this is only functioning as a pass-through gateway and not a control plane.  It may or may not create an IP address for the sensor when it wakes up.  I haven't had success using the gateway IP and the remote ID to proxy through in the experiments I conducted but I haven't tried with the setup you have.
* The battery powered "Luminance sensor" will be hard to read locally as they are optimized to limit power usage by shutting down except during updates to the TuyaCloud.  If you set up a continual loop to monitor for it to show up on the network (IP address or the gateway IP) and then run an async scan on it (like [this](https://github.com/jasonacox/tinytuya/blob/master/examples/async_send_receive.py)) you might catch it and get the datapoint locally.  Danger is that if you do catch it, it may "stay alive" sending updates but will quickly drain the battery.
* For battery/sensor Tuya devices, I would suggest using the cloud functions of TinyTuya to pull the state and send commands. It will allow the device and gateway to operate as planned but still give you an API to scrape the data (see [cloud.py](https://github.com/jasonacox/tinytuya/blob/master/examples/cloud.py) example).  I would be interested to know if that works for your setup. 

Thanks for pasting the details of your setup.  I know it will help anyone in the community looking at the same setup.

**Reply by renatomott** [reply]
Guys,
Any success controlling Orvibo devices using Tuya or Smart Life app?
